engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/runtime/AssertNode.java

```diff
@@ -0,0 +1,87 @@
+package org.enso.interpreter.node.expression.builtin.runtime;
+
+import com.oracle.truffle.api.Assumption;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.dsl.Suspend;
+import org.enso.interpreter.node.BaseNode;
+import org.enso.interpreter.node.callable.thunk.ThunkExecutorNode;
+import org.enso.interpreter.node.expression.builtin.meta.TypeOfNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.state.State;
+
+@BuiltinMethod(
+    type = "Runtime",
+    name = "assert_builtin",
+    description = "Asserts that the given condition is true",
+    autoRegister = false)
+public abstract class AssertNode extends Node {
+
+  public static AssertNode build() {
+    return AssertNodeGen.create();
+  }
+
+  public abstract Object execute(
+      VirtualFrame frame, State state, @Suspend Object action, Object msg);
+
+  protected Assumption getAssertionsDisabledAssumption() {
+    return EnsoContext.get(this).getAssertsDisabledAssumption();
+  }
+
+  @Specialization(assumptions = "getAssertionsDisabledAssumption()")
+  Object doAssertionsDisabled(VirtualFrame frame, State state, Object action, Object msg) {
+    return EnsoContext.get(this).getNothing();
+  }
+
+  @Specialization(replaces = "doAssertionsDisabled")
+  Object doAssertionsEnabled(
+      VirtualFrame frame,
+      State state,
+      Object action,
+      Object msg,
+      @Cached("create()") ThunkExecutorNode thunkExecutorNode,
+      @CachedLibrary(limit = "3") InteropLibrary interop,
```

Comment:
For some reason, `thunkExecutorNode` does not throw type error if you pass something non-executable. I will rather keep this *isExecutable* interop check here. This code is on slow path anyway.

Reply:
I just realized that `action` does not have to be *executable* at all, it can be just a constant. Explicit check for executability of `action` via interop removed in 8e8c192f7c8f8ee6fe1cd4a78bc90ff14647aaf2